Support building with Rust stable

Hello! 👋🏼 

I recently packaged `lazyjj` for Arch Linux (#40) - I had to use Rust nightly because of this feature:

```
error[E0554]: `#![feature]` may not be used on the stable release channel
 --> src/main.rs:1:1
  |
1 | #![feature(let_chains)]
  | ^^^^^^^^^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0554`.
error: could not compile `lazyjj` (bin "lazyjj") due to 1 previous error
```

On Arch we prefer to use the stable channel because it helps with [reproducibility](https://reproducible-builds.org/) of the packages.

Would you be interested in supporting Rust stable by removing that feature and refactoring the code? 🐻

--- a/src/app.rs
+++ b/src/app.rs
@@ -103,20 +103,20 @@ impl App<'_> {
                 }
                 ComponentInputResult::Handled => {}
                 ComponentInputResult::NotHandled => {
-                    if let Event::Key(key) = event
-                        && key.kind == event::KeyEventKind::Press
-                    {
-                        // Close
-                        if matches!(
-                            key.code,
-                            KeyCode::Char('y')
-                                | KeyCode::Char('n')
-                                | KeyCode::Char('o')
-                                | KeyCode::Enter
-                                | KeyCode::Char('q')
-                                | KeyCode::Esc
-                        ) {
-                            self.popup = None
+                    if let Event::Key(key) = event {
+                        if key.kind == event::KeyEventKind::Press {
+                            // Close
+                            if matches!(
+                                key.code,
+                                KeyCode::Char('y')
+                                    | KeyCode::Char('n')
+                                    | KeyCode::Char('o')
+                                    | KeyCode::Enter
+                                    | KeyCode::Char('q')
+                                    | KeyCode::Esc
+                            ) {
+                                self.popup = None
+                            }
                         }
                     }
                 }
@@ -131,27 +131,27 @@ impl App<'_> {
                 }
                 ComponentInputResult::Handled => {}
                 ComponentInputResult::NotHandled => {
-                    if let Event::Key(key) = event
-                        && key.kind == event::KeyEventKind::Press
-                    {
-                        // Close
-                        if key.code == KeyCode::Char('q')
-                            || (key.modifiers.contains(KeyModifiers::CONTROL)
-                                && (key.code == KeyCode::Char('c')))
-                            || key.code == KeyCode::Esc
-                        {
-                            return Ok(true);
-                        }
-                        //
-                        // Tab switching
-                        if let Some((_, tab)) = Tab::VALUES.iter().enumerate().find(|(i, _)| {
-                            key.code
-                                == KeyCode::Char(
-                                    char::from_digit((*i as u32) + 1u32, 10)
-                                        .expect("Tab index could not be converted to digit"),
-                                )
-                        }) {
-                            self.set_tab(commander, *tab)?;
+                    if let Event::Key(key) = event {
+                        if key.kind == event::KeyEventKind::Press {
+                            // Close
+                            if key.code == KeyCode::Char('q')
+                                || (key.modifiers.contains(KeyModifiers::CONTROL)
+                                    && (key.code == KeyCode::Char('c')))
+                                || key.code == KeyCode::Esc
+                            {
+                                return Ok(true);
+                            }
+                            //
+                            // Tab switching
+                            if let Some((_, tab)) = Tab::VALUES.iter().enumerate().find(|(i, _)| {
+                                key.code
+                                    == KeyCode::Char(
+                                        char::from_digit((*i as u32) + 1u32, 10)
+                                            .expect("Tab index could not be converted to digit"),
+                                    )
+                            }) {
+                                self.set_tab(commander, *tab)?;
+                            }
                         }
                     }
                 }

--- a/src/commander/branches.rs
+++ b/src/commander/branches.rs
@@ -40,11 +40,7 @@ fn parse_branch(text: &str) -> Option<Branch> {
         let name = captured.get(1);
         let remote = captured.get(2);
         let present = captured.get(3);
-
-        if let Some(name) = name
-            && let Some(remote) = remote
-            && let Some(present) = present
-        {
+        if let (Some(name), Some(remote), Some(present)) = (name, remote, present) {
             let remote = remote.as_str().to_owned();
             Some(Branch {
                 remote: if remote.is_empty() {

--- a/src/commander/log.rs
+++ b/src/commander/log.rs
@@ -54,16 +54,12 @@ fn parse_head(text: &str) -> Result<Head> {
     captured
         .as_ref()
         .map_or(Err(anyhow!(HeadParseError(text.to_owned()))), |captured| {
-            let change_id = captured.get(1);
-            let commit_id = captured.get(2);
-            let divergent = captured.get(3);
-            let immutable = captured.get(4);
-
-            if let Some(change_id) = change_id
-                && let Some(commit_id) = commit_id
-                && let Some(divergent) = divergent
-                && let Some(immutable) = immutable
-            {
+            if let (Some(change_id), Some(commit_id), Some(divergent), Some(immutable)) = (
+                captured.get(1),
+                captured.get(2),
+                captured.get(3),
+                captured.get(4),
+            ) {
                 Ok(Head {
                     change_id: ChangeId(change_id.as_str().to_string()),
                     commit_id: CommitId(commit_id.as_str().to_string()),

--- a/src/main.rs
+++ b/src/main.rs
@@ -1,5 +1,3 @@
-#![feature(let_chains)]
-
 extern crate lazy_static;
 extern crate thiserror;
 
@@ -54,10 +52,10 @@ fn main() -> Result<()> {
     };
 
     // Check that jj exists
-    if let Err(err) = Command::new("jj").arg("help").output()
-        && let ErrorKind::NotFound = err.kind()
-    {
-        bail!("jj command not found. Please make sure it is installed: https://martinvonz.github.io/jj/latest/install-and-setup");
+    if let Err(err) = Command::new("jj").arg("help").output() {
+        if err.kind() == ErrorKind::NotFound {
+            bail!("jj command not found. Please make sure it is installed: https://martinvonz.github.io/jj/latest/install-and-setup");
+        }
     }
 
     // Setup environment

--- a/src/ui/branch_set_popup.rs
+++ b/src/ui/branch_set_popup.rs
@@ -286,8 +286,10 @@ impl Component for BranchSetPopup<'_> {
                     self.on_creating();
                 }
                 KeyCode::Enter => {
-                    if let Some(index) = self.list_state.selected()
-                        && let Some(action) = self.options.get(index)
+                    if let Some(action) = self
+                        .list_state
+                        .selected()
+                        .and_then(|index| self.options.get(index))
                     {
                         match action {
                             BranchSetOption::CreateBranch => {

--- a/src/ui/branches_tab.rs
+++ b/src/ui/branches_tab.rs
@@ -205,75 +205,77 @@ impl Component for BranchesTab<'_> {
 
     fn update(&mut self, commander: &mut Commander) -> Result<Option<ComponentAction>> {
         // Check for popup action
-        if let Ok(res) = self.popup_rx.try_recv()
-            && res.1.unwrap_or(false)
-        {
-            match res.0 {
-                DELETE_BRANCH_POPUP_ID => {
-                    if let Some(delete) = self.delete.as_ref() {
-                        match commander.delete_branch(&delete.name) {
-                            Ok(_) => {
-                                self.refresh_branches(commander);
-                                let branches = Vec::new();
-                                let branches = self.branches_output.as_ref().unwrap_or(&branches);
-                                self.branch = branches.first().map(|branch| branch.to_owned());
-                                self.refresh_branch(commander);
-                            }
-                            Err(err) => {
-                                return Ok(Some(ComponentAction::SetPopup(Some(Box::new(
-                                    MessagePopup {
-                                        title: "Delete error".into(),
-                                        messages: err.to_string().into_text()?,
-                                    },
-                                )))));
-                            }
-                        }
-                    }
-                }
-                FORGET_BRANCH_POPUP_ID => {
-                    if let Some(forget) = self.forget.as_ref() {
-                        match commander.forget_branch(&forget.name) {
-                            Ok(_) => {
-                                self.refresh_branches(commander);
-                                let branches = Vec::new();
-                                let branches = self.branches_output.as_ref().unwrap_or(&branches);
-                                self.branch = branches.first().map(|branch| branch.to_owned());
-                                self.refresh_branch(commander);
-                            }
-                            Err(err) => {
-                                return Ok(Some(ComponentAction::SetPopup(Some(Box::new(
-                                    MessagePopup {
-                                        title: "Forget error".into(),
-                                        messages: err.to_string().into_text()?,
-                                    },
-                                )))));
+        if let Ok(res) = self.popup_rx.try_recv() {
+            if res.1.unwrap_or(false) {
+                match res.0 {
+                    DELETE_BRANCH_POPUP_ID => {
+                        if let Some(delete) = self.delete.as_ref() {
+                            match commander.delete_branch(&delete.name) {
+                                Ok(_) => {
+                                    self.refresh_branches(commander);
+                                    let branches = Vec::new();
+                                    let branches =
+                                        self.branches_output.as_ref().unwrap_or(&branches);
+                                    self.branch = branches.first().map(|branch| branch.to_owned());
+                                    self.refresh_branch(commander);
+                                }
+                                Err(err) => {
+                                    return Ok(Some(ComponentAction::SetPopup(Some(Box::new(
+                                        MessagePopup {
+                                            title: "Delete error".into(),
+                                            messages: err.to_string().into_text()?,
+                                        },
+                                    )))));
+                                }
                             }
                         }
                     }
-                }
-                NEW_POPUP_ID => {
-                    if let Some(BranchLine::Parsed { branch, .. }) = self.branch.as_ref() {
-                        commander.run_new(&branch.to_string())?;
-                        let head = commander.get_current_head()?;
-                        if self.describe_after_new {
-                            self.describe_after_new_change = Some(head.change_id);
-                            self.describe_after_new = false;
-                            let textarea = TextArea::default();
-                            self.describe_textarea = Some(textarea);
-                            return Ok(None);
-                        } else {
+                    FORGET_BRANCH_POPUP_ID => {
+                        if let Some(forget) = self.forget.as_ref() {
+                            match commander.forget_branch(&forget.name) {
+                                Ok(_) => {
+                                    self.refresh_branches(commander);
+                                    let branches = Vec::new();
+                                    let branches =
+                                        self.branches_output.as_ref().unwrap_or(&branches);
+                                    self.branch = branches.first().map(|branch| branch.to_owned());
+                                    self.refresh_branch(commander);
+                                }
+                                Err(err) => {
+                                    return Ok(Some(ComponentAction::SetPopup(Some(Box::new(
+                                        MessagePopup {
+                                            title: "Forget error".into(),
+                                            messages: err.to_string().into_text()?,
+                                        },
+                                    )))));
+                                }
+                            }
+                        }
+                    }
+                    NEW_POPUP_ID => {
+                        if let Some(BranchLine::Parsed { branch, .. }) = self.branch.as_ref() {
+                            commander.run_new(&branch.to_string())?;
+                            let head = commander.get_current_head()?;
+                            if self.describe_after_new {
+                                self.describe_after_new_change = Some(head.change_id);
+                                self.describe_after_new = false;
+                                let textarea = TextArea::default();
+                                self.describe_textarea = Some(textarea);
+                                return Ok(None);
+                            } else {
+                                return Ok(Some(ComponentAction::ViewLog(head)));
+                            }
+                        }
+                    }
+                    EDIT_POPUP_ID => {
+                        if let Some(BranchLine::Parsed { branch, .. }) = self.branch.as_ref() {
+                            commander.run_edit(&branch.to_string())?;
+                            let head = commander.get_current_head()?;
                             return Ok(Some(ComponentAction::ViewLog(head)));
                         }
                     }
+                    _ => {}
                 }
-                EDIT_POPUP_ID => {
-                    if let Some(BranchLine::Parsed { branch, .. }) = self.branch.as_ref() {
-                        commander.run_edit(&branch.to_string())?;
-                        let head = commander.get_current_head()?;
-                        return Ok(Some(ComponentAction::ViewLog(head)));
-                    }
-                }
-                _ => {}
             }
         }
 
@@ -333,15 +335,17 @@ impl Component for BranchesTab<'_> {
                 Err(err) => [
                     vec![Line::raw("Error getting branches").bold().fg(Color::Red)],
                     // TODO: Remove when jj 0.20 is released
-                    if let CommandError::Status(output, _) = err
-                        && output.contains("unexpected argument '-T' found")
-                    {
-                        vec![
-                            Line::raw(""),
-                            Line::raw("Please update jj to >0.18 for -T support to branches")
-                                .bold()
-                                .fg(Color::Red),
-                        ]
+                    if let CommandError::Status(output, _) = err {
+                        if output.contains("unexpected argument '-T' found") {
+                            vec![
+                                Line::raw(""),
+                                Line::raw("Please update jj to >0.18 for -T support to branches")
+                                    .bold()
+                                    .fg(Color::Red),
+                            ]
+                        } else {
+                            vec![]
+                        }
                     } else {
                         vec![]
                     },
@@ -660,9 +664,10 @@ impl Component for BranchesTab<'_> {
             return Ok(ComponentInputResult::Handled);
         }
 
-        if let Some(describe_textarea) = self.describe_textarea.as_mut()
-            && let Some(describe_after_new_change) = self.describe_after_new_change.as_ref()
-        {
+        if let (Some(describe_textarea), Some(describe_after_new_change)) = (
+            self.describe_textarea.as_mut(),
+            self.describe_after_new_change.as_ref(),
+        ) {
             if let Event::Key(key) = event {
                 match key.code {
                     KeyCode::Char('s') if key.modifiers.contains(KeyModifiers::CONTROL) => {
@@ -689,9 +694,10 @@ impl Component for BranchesTab<'_> {
             return Ok(ComponentInputResult::Handled);
         }
 
-        if let Event::Key(key) = event
-            && key.kind == KeyEventKind::Press
-        {
+        if let Event::Key(key) = event {
+            if key.kind == KeyEventKind::Press {
+                return Ok(ComponentInputResult::Handled);
+            }
             if self.popup.is_opened() {
                 if key.code == KeyCode::Char('q') || key.code == KeyCode::Esc {
                     self.popup = ConfirmDialogState::default();
@@ -793,66 +799,31 @@ impl Component for BranchesTab<'_> {
                 }
                 // TODO: Ask for confirmation?
                 KeyCode::Char('t') => {
-                    if let Some(BranchLine::Parsed { branch, .. }) = self.branch.as_ref()
-                        && branch.remote.is_some()
-                        && branch.present
-                    {
-                        commander.track_branch(branch)?;
-                        self.refresh_branches(commander);
-                        self.refresh_branch(commander);
+                    if let Some(BranchLine::Parsed { branch, .. }) = self.branch.as_ref() {
+                        if branch.remote.is_some() && branch.present {
+                            commander.track_branch(branch)?;
+                            self.refresh_branches(commander);
+                            self.refresh_branch(commander);
+                        }
                     }
                 }
                 KeyCode::Char('T') => {
-                    if let Some(BranchLine::Parsed { branch, .. }) = self.branch.as_ref()
-                        && branch.remote.is_some()
-                        && branch.present
-                    {
-                        commander.untrack_branch(branch)?;
-                        self.refresh_branches(commander);
-                        self.refresh_branch(commander);
+                    if let Some(BranchLine::Parsed { branch, .. }) = self.branch.as_ref() {
+                        if branch.remote.is_some() && branch.present {
+                            commander.untrack_branch(branch)?;
+                            self.refresh_branches(commander);
+                            self.refresh_branch(commander);
+                        }
                     }
                 }
                 KeyCode::Char('n') | KeyCode::Char('N') => {
-                    if let Some(BranchLine::Parsed { branch, .. }) = self.branch.as_ref()
-                        && branch.present
-                    {
-                        self.popup = ConfirmDialogState::new(
-                            NEW_POPUP_ID,
-                            Span::styled(" New ", Style::new().bold().cyan()),
-                            Text::from(vec![
-                                Line::from("Are you sure you want to create a new change?"),
-                                Line::from(format!("Branch: {}", branch)),
-                            ]),
-                        )
-                        .with_yes_button(ButtonLabel::YES.clone())
-                        .with_no_button(ButtonLabel::NO.clone())
-                        .with_listener(Some(self.popup_tx.clone()))
-                        .open();
-
-                        self.describe_after_new = key.code == KeyCode::Char('N');
-                    }
-                }
-                KeyCode::Char('e') => {
-                    if let Some(BranchLine::Parsed { branch, .. }) = self.branch.as_ref()
-                        && branch.present
-                    {
-                        if commander.check_revision_immutable(&branch.to_string())? {
-                            return Ok(ComponentInputResult::HandledAction(
-                                ComponentAction::SetPopup(Some(Box::new(MessagePopup {
-                                    title: "Edit".into(),
-                                    messages: vec![
-                                        "The change cannot be edited because it is immutable."
-                                            .into(),
-                                    ]
-                                    .into(),
-                                }))),
-                            ));
-                        } else {
+                    if let Some(BranchLine::Parsed { branch, .. }) = self.branch.as_ref() {
+                        if branch.present {
                             self.popup = ConfirmDialogState::new(
-                                EDIT_POPUP_ID,
-                                Span::styled(" Edit ", Style::new().bold().cyan()),
+                                NEW_POPUP_ID,
+                                Span::styled(" New ", Style::new().bold().cyan()),
                                 Text::from(vec![
-                                    Line::from("Are you sure you want to edit an existing change?"),
+                                    Line::from("Are you sure you want to create a new change?"),
                                     Line::from(format!("Branch: {}", branch)),
                                 ]),
                             )
@@ -860,16 +831,51 @@ impl Component for BranchesTab<'_> {
                             .with_no_button(ButtonLabel::NO.clone())
                             .with_listener(Some(self.popup_tx.clone()))
                             .open();
+
+                            self.describe_after_new = key.code == KeyCode::Char('N');
+                        }
+                    }
+                }
+                KeyCode::Char('e') => {
+                    if let Some(BranchLine::Parsed { branch, .. }) = self.branch.as_ref() {
+                        if branch.present {
+                            if commander.check_revision_immutable(&branch.to_string())? {
+                                return Ok(ComponentInputResult::HandledAction(
+                                    ComponentAction::SetPopup(Some(Box::new(MessagePopup {
+                                        title: "Edit".into(),
+                                        messages: vec![
+                                            "The change cannot be edited because it is immutable."
+                                                .into(),
+                                        ]
+                                        .into(),
+                                    }))),
+                                ));
+                            } else {
+                                self.popup = ConfirmDialogState::new(
+                                    EDIT_POPUP_ID,
+                                    Span::styled(" Edit ", Style::new().bold().cyan()),
+                                    Text::from(vec![
+                                        Line::from(
+                                            "Are you sure you want to edit an existing change?",
+                                        ),
+                                        Line::from(format!("Branch: {}", branch)),
+                                    ]),
+                                )
+                                .with_yes_button(ButtonLabel::YES.clone())
+                                .with_no_button(ButtonLabel::NO.clone())
+                                .with_listener(Some(self.popup_tx.clone()))
+                                .open();
+                            }
                         }
                     }
                 }
                 KeyCode::Enter => {
-                    if let Some(BranchLine::Parsed { branch, .. }) = self.branch.as_ref()
-                        && branch.present
-                    {
-                        return Ok(ComponentInputResult::HandledAction(
-                            ComponentAction::ViewLog(commander.get_branch_head(branch)?),
-                        ));
+                    if let Some(BranchLine::Parsed { branch, .. }) = self.branch.as_ref() {
+                        if branch.present {
+                            return Ok(ComponentInputResult::HandledAction(
+                                ComponentAction::ViewLog(commander.get_branch_head(branch)?),
+                            ));
+                        }
                     }
                 }
                 KeyCode::Char('h') | KeyCode::Char('?') => {

--- a/src/ui/branches_tab.rs
+++ b/src/ui/branches_tab.rs
@@ -695,7 +695,7 @@ impl Component for BranchesTab<'_> {
         }
 
         if let Event::Key(key) = event {
-            if key.kind == KeyEventKind::Press {
+            if key.kind != KeyEventKind::Press {
                 return Ok(ComponentInputResult::Handled);
             }
             if self.popup.is_opened() {

--- a/src/ui/command_log_tab.rs
+++ b/src/ui/command_log_tab.rs
@@ -43,8 +43,10 @@ impl CommandLogTag {
     pub fn get_output_lines<'a>(&self) -> Result<Vec<Line<'a>>> {
         let mut output_lines = vec![];
 
-        if let Some(selected_index) = self.commands_list_state.selected()
-            && let Some(command) = self.command_history.iter().rev().nth(selected_index)
+        if let Some(command) = self
+            .commands_list_state
+            .selected()
+            .and_then(|selected_index| self.command_history.iter().rev().nth(selected_index))
         {
             match command.output.clone().borrow() {
                 Ok(output) => {
@@ -186,9 +188,7 @@ impl Component for CommandLogTag {
                             },
                         );
 
-                    if let Some(selected_index) = self.commands_list_state.selected()
-                        && i == selected_index
-                    {
+                    if self.commands_list_state.selected() == Some(i) {
                         line = line.bg(self.config.highlight_color());
                     }
 
@@ -227,9 +227,11 @@ impl Component for CommandLogTag {
 
     #[allow(clippy::collapsible_if)]
     fn input(&mut self, _commander: &mut Commander, event: Event) -> Result<ComponentInputResult> {
-        if let Event::Key(key) = event
-            && key.kind == KeyEventKind::Press
-        {
+        if let Event::Key(key) = event {
+            if key.kind != KeyEventKind::Press {
+                return Ok(ComponentInputResult::Handled);
+            }
+
             if self.output_panel.input(key) {
                 return Ok(ComponentInputResult::Handled);
             }

--- a/src/ui/files_tab.rs
+++ b/src/ui/files_tab.rs
@@ -37,9 +37,7 @@ fn get_current_file_index(
     current_file: Option<&String>,
     files_output: Result<&Vec<File>, &CommandError>,
 ) -> Option<usize> {
-    if let Some(current_file) = current_file
-        && let Ok(files_output) = files_output
-    {
+    if let (Some(current_file), Ok(files_output)) = (current_file, files_output) {
         files_output.iter().position(|file| {
             file.path
                 .as_ref()
@@ -140,11 +138,11 @@ impl FilesTab {
                 None => files.first(),
             }
             .map(|x| x.to_owned());
-            if let Some(next_file) = next_file
-                && next_file.path.is_some()
-            {
-                self.file.clone_from(&next_file.path);
-                self.refresh_diff(commander)?;
+            if let Some(next_file) = next_file {
+                if next_file.path.is_some() {
+                    self.file.clone_from(&next_file.path);
+                    self.refresh_diff(commander)?;
+                }
             }
         }
         Ok(())
@@ -276,9 +274,11 @@ impl Component for FilesTab {
     }
 
     fn input(&mut self, commander: &mut Commander, event: Event) -> Result<ComponentInputResult> {
-        if let Event::Key(key) = event
-            && key.kind == KeyEventKind::Press
-        {
+        if let Event::Key(key) = event {
+            if key.kind != KeyEventKind::Press {
+                return Ok(ComponentInputResult::Handled);
+            }
+
             if self.diff_panel.input(key) {
                 return Ok(ComponentInputResult::Handled);
             }

--- a/src/ui/help_popup.rs
+++ b/src/ui/help_popup.rs
@@ -86,19 +86,20 @@ impl Component for HelpPopup {
         _commander: &mut crate::commander::Commander,
         event: crossterm::event::Event,
     ) -> anyhow::Result<crate::ComponentInputResult> {
-        if let Event::Key(key) = event
-            && key.kind == event::KeyEventKind::Press
-        {
-            match key.code {
-                KeyCode::Char('j') => {
-                    let max = self.left_items.len().max(self.right_items.len());
-                    self.scroll = (self.scroll + 1).min(max.saturating_sub(self.height as usize));
+        if let Event::Key(key) = event {
+            if key.kind == event::KeyEventKind::Press {
+                match key.code {
+                    KeyCode::Char('j') => {
+                        let max = self.left_items.len().max(self.right_items.len());
+                        self.scroll =
+                            (self.scroll + 1).min(max.saturating_sub(self.height as usize));
+                    }
+                    KeyCode::Char('k') => self.scroll = self.scroll.saturating_sub(1),
+                    _ => return Ok(ComponentInputResult::NotHandled),
                 }
-                KeyCode::Char('k') => self.scroll = self.scroll.saturating_sub(1),
-                _ => return Ok(ComponentInputResult::NotHandled),
-            }
 
-            return Ok(ComponentInputResult::Handled);
+                return Ok(ComponentInputResult::Handled);
+            }
         }
 
         Ok(ComponentInputResult::NotHandled)

--- a/src/ui/log_tab.rs
+++ b/src/ui/log_tab.rs
@@ -171,50 +171,48 @@ impl Component for LogTab<'_> {
         }
 
         // Check for popup action
-        if let Ok(res) = self.popup_rx.try_recv()
-            && res.1.unwrap_or(false)
-        {
-            match res.0 {
-                NEW_POPUP_ID => {
-                    commander.run_new(self.head.commit_id.as_str())?;
-                    self.head = commander.get_current_head()?;
-                    self.refresh_log_output(commander);
-                    self.refresh_head_output(commander);
-                    if self.describe_after_new {
-                        self.describe_after_new = false;
-                        let textarea = TextArea::default();
-                        self.describe_textarea = Some(textarea);
-                    }
-                    return Ok(Some(ComponentAction::ChangeHead(self.head.clone())));
-                }
-                EDIT_POPUP_ID => {
-                    commander.run_edit(self.head.commit_id.as_str())?;
-                    self.refresh_log_output(commander);
-                    self.refresh_head_output(commander);
-                    return Ok(Some(ComponentAction::ChangeHead(self.head.clone())));
-                }
-                ABANDON_POPUP_ID => {
-                    if self.head == commander.get_current_head()? {
-                        commander.run_abandon(&self.head.commit_id)?;
-                        self.refresh_log_output(commander);
+        if let Ok(res) = self.popup_rx.try_recv() {
+            if res.1.unwrap_or(false) {
+                match res.0 {
+                    NEW_POPUP_ID => {
+                        commander.run_new(self.head.commit_id.as_str())?;
                         self.head = commander.get_current_head()?;
+                        self.refresh_log_output(commander);
+                        self.refresh_head_output(commander);
+                        if self.describe_after_new {
+                            self.describe_after_new = false;
+                            let textarea = TextArea::default();
+                            self.describe_textarea = Some(textarea);
+                        }
+                        return Ok(Some(ComponentAction::ChangeHead(self.head.clone())));
+                    }
+                    EDIT_POPUP_ID => {
+                        commander.run_edit(self.head.commit_id.as_str())?;
+                        self.refresh_log_output(commander);
                         self.refresh_head_output(commander);
                         return Ok(Some(ComponentAction::ChangeHead(self.head.clone())));
-                    } else {
-                        let head_parent = commander.get_commit_parent(&self.head.commit_id)?;
-                        commander.run_abandon(&self.head.commit_id)?;
-                        self.refresh_log_output(commander);
-                        self.head = head_parent;
-                        self.refresh_head_output(commander);
                     }
+                    ABANDON_POPUP_ID => {
+                        if self.head == commander.get_current_head()? {
+                            commander.run_abandon(&self.head.commit_id)?;
+                            self.refresh_log_output(commander);
+                            self.head = commander.get_current_head()?;
+                            self.refresh_head_output(commander);
+                            return Ok(Some(ComponentAction::ChangeHead(self.head.clone())));
+                        } else {
+                            let head_parent = commander.get_commit_parent(&self.head.commit_id)?;
+                            commander.run_abandon(&self.head.commit_id)?;
+                            self.refresh_log_output(commander);
+                            self.head = head_parent;
+                            self.refresh_head_output(commander);
+                        }
+                    }
+                    _ => {}
                 }
-                _ => {}
             }
         }
 
-        if let Ok(res) = self.branch_set_popup_rx.try_recv()
-            && res
-        {
+        if let Ok(true) = self.branch_set_popup_rx.try_recv() {
             self.refresh_log_output(commander);
             self.refresh_head_output(commander)
         }
@@ -452,9 +450,11 @@ impl Component for LogTab<'_> {
             return Ok(ComponentInputResult::Handled);
         }
 
-        if let Event::Key(key) = event
-            && key.kind == KeyEventKind::Press
-        {
+        if let Event::Key(key) = event {
+            if key.kind != KeyEventKind::Press {
+                return Ok(ComponentInputResult::Handled);
+            }
+
             if self.popup.is_opened() {
                 if key.code == KeyCode::Char('q') || key.code == KeyCode::Esc {
                     self.popup = ConfirmDialogState::default();

